python/setup.py

```diff
@@ -101,91 +116,151 @@
 extras["all"] = list(set(chain.from_iterable(extras.values())))
 
 
-class build_ext(_build_ext.build_ext):
-    def run(self):
-        # Note: We are passing in sys.executable so that we use the same
-        # version of Python to build packages inside the build.sh script. Note
-        # that certain flags will not be passed along such as --user or sudo.
-        # TODO(rkn): Fix this.
-        command = ["../build.sh", "-p", sys.executable]
-        if sys.platform == "win32" and command[0].lower().endswith(".sh"):
-            # We can't run .sh files directly in Windows, so find a shell.
-            # Don't use "bash" instead of "sh", because that might run the Bash
-            # from WSL! (We want MSYS2's Bash, which is also sh by default.)
-            shell = os.getenv("BAZEL_SH", "sh")  # NOT "bash"! (see above)
-            command.insert(0, shell)
-        if build_java:
-            # Also build binaries for Java if the above env variable exists.
-            command += ["-l", "python,java"]
-        subprocess.check_call(command)
-
-        # We also need to install pickle5 along with Ray, so make sure that the
-        # relevant non-Python pickle5 files get copied.
-        pickle5_files = self.walk_directory("./ray/pickle5_files/pickle5")
-
-        thirdparty_files = self.walk_directory("./ray/thirdparty_files")
-
-        files_to_include = ray_files + pickle5_files + thirdparty_files
-
-        # Copy over the autogenerated protobuf Python bindings.
-        for directory in generated_python_directories:
-            for filename in os.listdir(directory):
-                if filename[-3:] == ".py":
-                    files_to_include.append(os.path.join(directory, filename))
-
-        for filename in files_to_include:
-            self.move_file(filename)
-
-        # Try to copy over the optional files.
-        for filename in optional_ray_files:
+def is_native_windows_or_msys():
+    """Check to see if we are running on native Windows,
+    but NOT WSL (which is seen as Linux)."""
+    return sys.platform == "msys" or sys.platform == "win32"
+
+
+def is_invalid_windows_platform():
+    # 'GCC' check is how you detect MinGW:
+    # https://github.com/msys2/MINGW-packages/blob/abd06ca92d876b9db05dd65f27d71c4ebe2673a9/mingw-w64-python2/0410-MINGW-build-extensions-with-GCC.patch#L53
+    platform = sys.platform
+    ver = sys.version
+    return platform == "msys" or (platform == "win32" and ver and "GCC" in ver)
+
+
+def download(url):
+    try:
+        result = urllib.request.urlopen(url).read()
+    except urllib.error.URLError:
+        # This fallback is necessary on Python 3.5 on macOS due to TLS 1.2.
+        curl_args = ["curl", "-s", "-L", "-f", "-o", "-", url]
+        result = subprocess.check_output(curl_args)
+    return result
+
+
+def download_pickle5(pickle5_dir):
```

Comment:
Can you leave a high-level comment about what this is doing? e.g., what is the path it'll be downloaded to (the string manipulation logic is very hard to follow)

Reply:
Sure